packages/report-server/src/reportBuilder/output/functions/matrix/matrix.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { Matrix, Row } from '../../../types';
+import { MatrixBuilder } from './matrixBuilder';
+import { MatrixColumnParams, MatrixParams } from './types';
+
+const matrix = (rows: Row[], params: MatrixParams): Matrix => {
+  const matrixBuilder = new MatrixBuilder(rows, params);
+  matrixBuilder.buildBaseColumns().buildBaseRows();
+  return matrixBuilder.getMatrixData();
+};
+
+const buildParams = (params: unknown): MatrixParams => {
+  if (typeof params !== 'object' || params === null) {
+    throw new Error(`Expected params object but got ${params}`);
+  }
+
+  const getPrefixColumns = (columns: unknown): MatrixColumnParams => {
+    if (columns === '*' || columns === undefined) {
+      return '*';
+    }
+    if (!Array.isArray(columns)) {
+      throw new Error(`Expected columns as string array or '*' but got ${columns}`);
+    }
+    if (columns.length === 0) {
+      return '*';
+    }
+    columns.forEach(column => {
+      if (typeof column !== 'string') {
+        throw new Error(`Expected columns as string array or '*' but got ${columns}`);
+      }
+    });
+    return columns;
+  };
+  const { columns, categoryField, rowField } = params;
+
+  const prefixColumns = getPrefixColumns(columns);
+
+  if (typeof rowField !== 'string') {
+    throw new Error(`Expected rowField as string but got ${rowField}`);
+  }
```

Comment:
I suppose we ought to be checking that they haven't included the rowField in the columns, or categories, and vice versa? With something like this:
```suggestion
  }
  
  if (Array.isArray(prefixColumns) && prefixColumns.includes(rowField)) {
    throw new Error(`rowField: ${rowField} is already defined as a column`);
  }
  
```

This is fairly optional tho imo, as that's just a silly configuration mistake 😝

Reply:
Yes I agree more input defenders is better!